Sign up

rajeh1032 opened the pull request:
<!--
  🛠️ For Work In Progress, please use Draft PRs:
  https://github.blog/2019-02-14-introducing-draft-pull-requests/

  ✅ Before submitting:
  - Follow your project's contributing guide
  - Ensure code compiles and passes all tests
  - Use descriptive commit messages
  - Keep PRs small and focused
-->

## 📌 What type of PR is this? (check all applicable)

- [✅ ] ✨ Feature
- [ ] 🐛 Bug Fix
- [✅ ] 🧹 Refactor
- [ ] 🚀 Performance Optimization
- [ ] 🧾 Documentation Update
- [ ] ♿ Accessibility Fix
- [ ] 🔧 CI/CD or Config Change

---

## 📝 Description

_Describe the goal of this PR and what was changed._
Feat : signUp With Api Integration
---

## 🧩 Related Issues / Tickets

- Closes #
- Fixes #
- Related to #

---

## 🧪 QA Instructions

_How to test the changes manually?_

- [✅ ] Run the app on Android
- [ ] Run the app on iOS
- [ ] Check specific feature/page:
  - [ ] ...
- [ ] Verify responsiveness on different screen sizes

---

## 📸 Screenshots or Recordings (if UI changes)

_Add screenshots or screen recordings to demonstrate your changes._
<img width="559" height="1207" alt="Screenshot 2025-08-21 153150" src="https://github.com/user-attachments/assets/a2f0dff4-5597-4ca1-8833-c26bd7ebc89d" />
<img width="558" height="1169" alt="Screenshot 2025-08-21 153140" src="https://github.com/user-attachments/assets/223f64b2-8a00-45c9-a77a-f60642fc1446" />
<img width="566" height="1200" alt="Screenshot 2025-08-21 153118" src="https://github.com/user-attachments/assets/4b4065c0-afce-4863-8a39-7e8c17499267" />


---

## ♿ UI Accessibility Checklist (if applicable)

- [ ] Semantic widget structure used (`Semantics`, `Accessible`, etc.)
- [ ] Supports screen readers
- [ ] Good color contrast
- [ ] Tested with TalkBack/VoiceOver

---

## 🧪 Tests

- [ ✅] Unit tests added
- [✅ ] Widget tests added
- ✅[ ] Integration tests added
- [ ] No tests needed (explain why):

---

## 🛠 Post-Deployment Tasks (Optional)

_Any migrations, scripts, or setup tasks?_

---

## 🎬 Fun GIF (Optional)

_What GIF best represents this PR?_

![fun gif](https://media.giphy.com/media/xT0xeJpnrWC4XWblEk/giphy.gif)

rajeh1032 committed:
feat: Implement Sign Up feature with form validation and API integration

- Added SignUpState to manage loading, success, and error states.
- Created UI components for sign-up including email, password, name, phone, and gender fields.
- Implemented BuildElevatedButton for handling loading states.
- Developed SignUpForm to manage user input and submission.
- Integrated API service for sign-up functionality with error handling.
- Added localization support for various UI texts.
- Created SecureStorage for managing secure data storage.
- Implemented unit tests for AuthRemoteDataSource to ensure API integration works as expected.

rajeh1032 committed:
feat: Refactor Sign Up feature by restructuring view model and updating imports

Eng-M0ataz commented on lib/Feature/auth/api/client/api_service.dart:
```diff
@@ -1,11 +1,18 @@
   factory ApiServices(Dio dio) = _ApiServices;
+
+  @POST(ApiConstants.signUp)
+  Future<SignUpResponseDto> signUp(@Body() SignUpRequestDto signUpRequest);
```
Rename SignUpRequestDto to SignUpRequestModel or SignUpUserInputModel

Eng-M0ataz commented on lib/Feature/auth/api/client/api_service.g.dart:
don't upload Generated Files to git hub

Eng-M0ataz commented on lib/Feature/auth/api/dataSources/auth_remote_data_source_impl.dart:
```diff
@@ -1,5 +1,36 @@
 @Injectable(as: AuthRemoteDataSource)
-class AuthRemoteDataSourceImpl implements AuthRemoteDataSource {}
+class AuthRemoteDataSourceImpl implements AuthRemoteDataSource {
+  ApiServices _apiServices;
```
use final here

Eng-M0ataz commented on lib/Feature/auth/api/dataSources/auth_remote_data_source_impl.dart:
```diff
@@ -1,5 +1,36 @@
+    try {
+      var response =
+          await _apiServices.signUp(SignUpRequestDto.fromDomain(signUpRequest));
+      return ApiSuccessResult(data: response.toEntity());
```
ApiSuccessResult of type SignUpResponseEntity  you need to add this  instead of ApiSuccessResult

rajeh1032 replied:
what do u mean "ApiSuccessResult of type SignUpResponseEntity "??

Eng-M0ataz replied:
<img width="381" height="47" alt="image" src="https://github.com/user-attachments/assets/a62fef06-3a67-4109-94e6-59d8f62f6fe0" />

Eng-M0ataz commented on lib/Feature/auth/api/dataSources/auth_remote_data_source_impl.dart:
```diff
@@ -1,5 +1,36 @@
+          await _apiServices.signUp(SignUpRequestDto.fromDomain(signUpRequest));
+      return ApiSuccessResult(data: response.toEntity());
+    } on DioException catch (dioError) {
+      // Handle DioException specifically to extract proper error messages
```
remove commit here

Eng-M0ataz commented on lib/Feature/auth/api/dataSources/auth_remote_data_source_impl.dart:
```diff
@@ -1,5 +1,36 @@
+    } catch (e) {
+      return ApiErrorResult<SignUpResponseEntity>(
+          failure: ServerFailure(
+              errorMessage: "Unexpected error occurred: ${e.toString()}"));
```
remove this string  "Unexpected error occurred because of any error happen will show you the error Message

Eng-M0ataz commented on lib/Feature/auth/api/models/response/sign_in_response_dto.g.dart:
don't upload generated files

Eng-M0ataz commented on lib/Feature/auth/api/models/response/sign_up_response_dto.dart:
```diff
@@ -0,0 +1,22 @@
+import 'package:flower_e_commerce_app/Feature/auth/api/models/user_dto.dart';
+import 'package:json_annotation/json_annotation.dart';
+
```
remove this blank space

Eng-M0ataz commented on lib/Feature/auth/api/models/response/sign_up_response_dto.dart:
```diff
@@ -0,0 +1,22 @@
+
+@JsonSerializable()
+class SignUpResponseDto {
+  final String message;
```
make all variables nullable be in safe side always

rajeh1032 replied:
but we handle if api not response

Eng-M0ataz commented on lib/Feature/auth/api/models/response/sign_up_response_dto.dart:
```diff
@@ -0,0 +1,22 @@
+class SignUpResponseDto {
+  final String message;
+  final UserDto user;
+  final String token;
```
always use @jsonKey() on all variables   to make sure the same name of json parm

rajeh1032 replied:
ok i will use it , but the name of message ,  user , token -- is identical for api response

Eng-M0ataz replied:
yea your right in this point but but if you used it at once make it for all in case backEnd changes any parm name you will only change the  jsonkey not the parm name ^^

Eng-M0ataz commented on lib/Feature/auth/api/models/sign_up_request_dto.dart:
```diff
@@ -0,0 +1,56 @@
+    required this.gender,
+  });
+
+  /// From Domain Entity
```
remove this

Eng-M0ataz commented on lib/Feature/auth/api/models/sign_up_request_dto.g.dart:
don't upload generated files

Eng-M0ataz commented on lib/Feature/auth/api/models/user_dto.dart:
```diff
@@ -0,0 +1,38 @@
+
+@JsonSerializable()
+class UserDto {
+  final String firstName;
```
use jsonKey() and make all parm nullable

Eng-M0ataz commented on lib/Feature/auth/api/models/user_dto.g.dart:
don't upload generated files

Eng-M0ataz commented on lib/Feature/auth/domain/useCases/sign_up_use_case.dart:
```diff
@@ -0,0 +1,15 @@
+
+@injectable
+class SignUpUseCase {
+  final AuthRepo authRepo;
```
make authRepo filed private

Eng-M0ataz commented on lib/Feature/auth/presentation/viewModel/events/sign_up_event.dart:
```diff
@@ -0,0 +1,21 @@
+sealed class SignUpEvent {}
+
+class SignUpSubmitEvent extends SignUpEvent {
+  final String firstName;
```
remove all this parm and access them from cubit controllers

Eng-M0ataz commented on lib/Feature/auth/presentation/viewModel/viewModel/signup_view_model.dart:
```diff
@@ -0,0 +1,121 @@
+
+@injectable
+class SignupViewModel extends Cubit<SignUpState> {
+  final SignUpUseCase signUpUseCase;
```
make this object private

Eng-M0ataz commented on lib/Feature/auth/presentation/widgets/signUp/build_first_and_last_name_field.dart:
```diff
@@ -0,0 +1,54 @@
+  Widget build(BuildContext context) {
+    return Row(
+      children: [
+        Expanded(
```
you need to use here entranced high widget to make all fields  with same size

rajeh1032 replied:
ok, but i think in this case we dont need "IntrinsicHeight Widget " Because it make ui slow and not have best performance

Eng-M0ataz replied:
yea your right this widget use it in cuation but in this case we need it to make ui response perfect because if the user focus on text field and then unfocused it error message will appear under this field then its high will change then the two textfield in the row will not in the same line so we forced to use this widget here and will not effect the performance of the app because we use it in one place

Eng-M0ataz commented on lib/Feature/auth/presentation/widgets/signUp/build_first_and_last_name_field.dart:
```diff
@@ -0,0 +1,54 @@
+          ),
+          validator: Validations.validateName,
+        )),
+        const SizedBox(width: 8),
```
use AppSizes class

Eng-M0ataz commented on lib/Feature/auth/presentation/widgets/signUp/build_gender_filed.dart:
```diff
@@ -0,0 +1,72 @@
+    return Row(
+      children: [
+        Text(
+          'gender'.tr(),
```
why you did't use here app Local Keys ?

Eng-M0ataz commented on lib/Feature/auth/presentation/widgets/signUp/build_gender_filed.dart:
```diff
@@ -0,0 +1,72 @@
+          'gender'.tr(),
+          style: AppThemeLight.lightTheme.textTheme.titleLarge,
+        ),
+        const SizedBox(width: 20),
```
use AppSizes class

Eng-M0ataz commented on lib/Feature/auth/presentation/widgets/signUp/build_gender_filed.dart:
```diff
@@ -0,0 +1,72 @@
+      children: [
+        Text(
+          'gender'.tr(),
+          style: AppThemeLight.lightTheme.textTheme.titleLarge,
```
use Theme.of(context).textTheme.titleLarge

Eng-M0ataz commented on lib/Feature/auth/presentation/widgets/signUp/build_gender_filed.dart:
```diff
@@ -0,0 +1,72 @@
+        ),
+        const SizedBox(width: 20),
+
+        // Female Radio Button
```
remove this

Eng-M0ataz commented on lib/Feature/auth/presentation/widgets/signUp/build_gender_filed.dart:
```diff
@@ -0,0 +1,72 @@
+          mainAxisSize: MainAxisSize.min,
+          children: [
+            Radio<String>(
+              value: 'female'.tr(),
```
same as last commit

Eng-M0ataz commented on lib/Feature/auth/presentation/widgets/signUp/build_gender_filed.dart:
```diff
@@ -0,0 +1,72 @@
+        const SizedBox(width: 20),
+
+        // Male Radio Button
+        Row(
```
make this row as reusable component and use them

Eng-M0ataz commented on lib/Feature/auth/presentation/widgets/signUp/build_gender_filed.dart:
```diff
@@ -0,0 +1,72 @@
+              activeColor: AppColorsLight.pink,
+            ),
+            Text(
+              'female'.tr(),
```
AppLocalKeys

Eng-M0ataz commented on lib/Feature/auth/presentation/widgets/signUp/build_navigation_text.dart:
```diff
@@ -0,0 +1,42 @@
+            minimumSize: const Size(0, 0),
+          ),
+          onPressed: () {
+            Navigator.pushNamed(context, routeName);
```
use  navigation extension class here like this context.pushNamed

Eng-M0ataz commented on lib/Feature/auth/presentation/widgets/signUp/build_navigation_text.dart:
```diff
@@ -0,0 +1,42 @@
+          },
+          child: Text(
+            secondText,
+            style: AppThemeLight.lightTheme.textTheme.headlineLarge!.copyWith(
```
check this

Eng-M0ataz commented on lib/Feature/auth/presentation/widgets/signUp/build_password_and_confirm_field.dart:
```diff
@@ -0,0 +1,54 @@
+          textInputAction: TextInputAction.next,
+          autofillHints: const [AutofillHints.password],
+          decoration: InputDecoration(
+            labelText: "password_label".tr(),
```
check here

Eng-M0ataz commented on lib/Feature/auth/presentation/widgets/signUp/build_password_and_confirm_field.dart:
```diff
@@ -0,0 +1,54 @@
+  Widget build(BuildContext context) {
+    return Row(
+      children: [
+        Expanded(
```
check here

Eng-M0ataz commented on lib/Feature/auth/presentation/widgets/signUp/build_password_and_confirm_field.dart:
```diff
@@ -0,0 +1,54 @@
+          ),
+          validator: Validations.validatePassword,
+        )),
+        SizedBox(width: 8),
```
check here

Eng-M0ataz commented on lib/Feature/auth/presentation/widgets/signUp/build_password_and_confirm_field.dart:
```diff
@@ -0,0 +1,54 @@
+          textInputAction: TextInputAction.next,
+          autofillHints: const [AutofillHints.password],
+          decoration: InputDecoration(
+            labelText: "confirm_password_label".tr(),
```
check here

Eng-M0ataz commented on lib/Feature/auth/presentation/widgets/signUp/build_password_and_confirm_field.dart:
```diff
@@ -0,0 +1,54 @@
+          decoration: InputDecoration(
+            labelText: "confirm_password_label".tr(),
+            hintText: "confirm_password_hint".tr(),
+            labelStyle: AppThemeLight.lightTheme.textTheme.titleMedium,
```
check here

Eng-M0ataz commented on lib/Feature/auth/presentation/widgets/signUp/build_phone_field.dart:
```diff
@@ -0,0 +1,27 @@
+      textInputAction: TextInputAction.next,
+      autofillHints: const [AutofillHints.telephoneNumber],
+      decoration: InputDecoration(
+        labelText: "phone_number_label".tr(),
```
check here

Eng-M0ataz commented on lib/Feature/auth/presentation/widgets/signUp/build_phone_field.dart:
```diff
@@ -0,0 +1,27 @@
+        labelText: "phone_number_label".tr(),
+        hintText: "phone_number_hint".tr(),
+        labelStyle: AppThemeLight.lightTheme.textTheme.titleMedium,
+        hintStyle: AppThemeLight.lightTheme.textTheme.bodyMedium,
```
check here

Eng-M0ataz commented on lib/Feature/auth/presentation/widgets/signUp/form/sign_up_form.dart:
```diff
@@ -0,0 +1,151 @@
+      if (state.isSuccess && !state.isLoading) {
+        DialogUtils.showMessage(
+          context: context,
+          message: 'Account created successfully!',
```
use translations here

Eng-M0ataz commented on lib/Feature/auth/presentation/widgets/signUp/form/sign_up_form.dart:
```diff
@@ -0,0 +1,151 @@
+
+      return Scaffold(
+          appBar: AppBar(
+            leadingWidth: 120,
```
use app sizes

Eng-M0ataz commented on lib/Feature/auth/presentation/widgets/signUp/form/sign_up_form.dart:
```diff
@@ -0,0 +1,151 @@
+          appBar: AppBar(
+            leadingWidth: 120,
+            leading: CustomBackButton(
+              title: "sign_up".tr(),
```
check all pervious comments and apply all fixes here

Eng-M0ataz commented on lib/Feature/auth/presentation/widgets/signUp/form/sign_up_form.dart:
```diff
@@ -0,0 +1,151 @@
+                          selectedGender: viewModel.selectedGender,
+                          onChanged: (String gender) {
+                            viewModel.updateGender(gender);
+                            setState(() {});
```
when you use state here and the only change was in gender widget now the all screen will rebuild and this will effect app performance

rajeh1032 replied:
because i dont use emit state to handle this case 🙂

Eng-M0ataz replied:
yea i know and i told you in radio listtile widget make it custom componant and make the widget statfule widget and the changes only changes in this widget and remove set state from the main form or you can make custom widget and use value notifier

Eng-M0ataz commented on lib/core/Config/Theme/app_colors.dart:
```diff
@@ -45,4 +45,16 @@ abstract class AppColorsLight {
     90: Color(0xFF535353),
     100: Color(0xFF323232),
   });
+  static const MaterialColor blue = MaterialColor(0xFF2196F3, <int, Color>{
```
why you add blue here ?

rajeh1032 replied:
<img width="543" height="146" alt="image" src="https://github.com/user-attachments/assets/31d8003c-0445-4f15-bb3f-09bb7a261113" />
<img width="550" height="153" alt="image" src="https://github.com/user-attachments/assets/86082970-7aff-4637-a9e5-6895dfa8b749" />

Eng-M0ataz replied:
okay this is awesome idea if the user was a male  and he dont  like pink but blue with pink don't match just remove it and make the radio button match the design

Eng-M0ataz commented on lib/core/Di/di.config.dart:
same last commit

Eng-M0ataz commented on lib/core/Widgets/custom_app_bar.dart:
```diff
@@ -14,9 +14,11 @@ class CustomBackButton extends StatelessWidget {
           onTap: onTap ?? context.pop,
-          child: Icon(Icons.arrow_back_ios, size: 20),
+          child: const Padding(
+            padding: EdgeInsets.only(left: 12), // هنا المسافة
```
remove comment  and use app sizes

Eng-M0ataz commented on lib/core/localization/codegen_loader.g.dart:
dont upload generated files

Eng-M0ataz commented on assets/translations/ar.json:
```diff
@@ -1 +1,21 @@
+  "register_button": "تسجيل",
+  "reset_password_title": "إعادة تعيين كلمة المرور",
+  "sign_up": "انشاء حساب",
+  "dog": "كلب"
```
what is this ?

rajeh1032 replied:
sorry it was for test

Eng-M0ataz replied:
yea i know ^^

Eng-M0ataz commented on lib/core/localization/locale_keys.g.dart:
dont upload generated files

Eng-M0ataz commented on lib/core/utils/Constants/api_constants.dart:
```diff
@@ -5,4 +5,8 @@ abstract class ApiConstants {
   static const String authorization = 'Authorization';
+
+
+  //Auth
```
remove this comment

rajeh1032 replied:
<img width="388" height="286" alt="image" src="https://github.com/user-attachments/assets/b5a82b91-4d8c-426b-b7ab-cb58fb701176" />



it for readability  like this photo

Eng-M0ataz replied:
oky your right  because  app sizes use comments but i only make this to make all team understand the file and  when we finish the app i will remove all that because app sizes not need this comments  ^^

Eng-M0ataz commented on lib/generated/locale_keys.g.dart:
dont upload generated files

Eng-M0ataz commented on lib/generated/codegen_loader.g.dart:
dont upload generated files

Eng-M0ataz commented on lib/Feature/auth/presentation/widgets/signUp/build_elevated_button.dart:
why you dont use our class Custom Elvated button ???????????????????????????????????????????????????

rajeh1032 replied:
<img width="311" height="260" alt="image" src="https://github.com/user-attachments/assets/9dac7ad9-0ba7-41ae-961c-d9ea65e1da13" />
see image pls!!

rajeh1032 replied:
i separated it in custom widget to be easy to read like all

Eng-M0ataz replied:
no need to do this widget customElvatedButton is already readable

rajeh1032 replied:
ok i will remove it

Eng-M0ataz commented on lib/core/Errors/failure.dart:
you modified this class and i see no logic for your modifications

rajeh1032 replied:
i handled  error response like 409 .. 500 ..

Eng-M0ataz replied:
no need to handle 409 because 409 response only return when api  responded example when you call sign up end point and this user already Registered will return map contains message or error with status code and i already handled it  in the class no need to add any modification on this file return it to pure and if yo not understand i can join you a meeting and explain it to you ^^

Eng-M0ataz replied:
and about 500 i already handle this case is original class

Eng-M0ataz commented on lib/core/Widgets/custom_app_bar.dart:
```diff
@@ -14,9 +14,11 @@ class CustomBackButton extends StatelessWidget {
         GestureDetector(
           onTap: onTap ?? context.pop,
-          child: Icon(Icons.arrow_back_ios, size: 20),
+          child: const Padding(
```
why you used padding here you can use padding on your signUp Screen ??????????

rajeh1032 replied:
in tomorrow meeting i will see the back button without padding and with padding the button doesn't match figma design without padding

Eng-M0ataz replied:
no problem if the space between the items on row dont match you can add size box to controll it like you want but dont give it padding here

rajeh1032 committed:
Refactor : Refactor sign-up flow and improve user data handling

- Added generated files to .gitignore for cleaner repository.
- Removed unused translations from Arabic JSON file.
- Updated English translations to include success message for account creation.
- Refactored AuthRemoteDataSourceImpl to use final ApiServices and improved error handling.
- Enhanced SignUpResponseDtoMapper to handle null values and provide default UserEntity.
- Updated UserDtoMapper to ensure all fields are nullable with default values.
- Deleted obsolete sign_in_response_dto.g.dart file.
- Refactored SignUpResponseDto to use nullable fields and updated JSON serialization.
- Cleaned up SignUpRequestDto by removing unnecessary comments.
- Updated UserDto to make fields nullable and added JSON annotations.
- Removed BuildElevatedButton widget as it was no longer needed.
- Updated BuildEmailField, BuildFirstAndLastNameField, BuildGenderField, BuildNavigationText, BuildPasswordAndConfirmField, and BuildPhoneField to use Theme for styling.
- Refactored SignUpForm to use new padding constants and removed unnecessary widgets.
- Updated dependency injection configuration to include SignupViewModel.
- Improved CustomBackButton padding using constants.
- Added gender constants to AppConstants for better maintainability.
- Introduced new size constants for consistent padding and margins across the app.

Eng-M0ataz commented on lib/Feature/auth/presentation/widgets/signUp/build_email_field.dart:
```diff
@@ -0,0 +1,26 @@
+      textInputAction: TextInputAction.next,
+      autofillHints: const [AutofillHints.email],
+      decoration: InputDecoration(
+        labelText: 'email_label'.tr(),
```
like before don,t use strings direct  use app localKeys

rajeh1032 committed:
feat: Update sign-up fields layout for improved readability and consistency

rajeh1032 committed:
refactor: Refactor sign-up request model and update localization keys

- Replaced SignUpRequestDto with SignUpRequestModel in API service and related files.
- Updated localization keys for gender labels, success messages, and form field labels.
- Enhanced the sign-up form to utilize new localization keys for better internationalization support.
- Removed deprecated SignUpRequestDto and its generated files.
- Improved error handling in ServerFailure class to extract error messages more effectively.
- Added new spacing constants for better layout management in the sign-up form.

rajeh1032 committed:
feat: Add selectedGender to SignUpState and update related methods

rajeh1032 committed:
feat: Update selectedGender to use AppConstants and refactor SignUpForm to StatelessWidget
